### PR TITLE
ADSI Module Improvements

### PR DESCRIPTION
**Pull request by stufus**
# Overview

This PR is for improvements to the ADSI module which effectively retrieves domain information using LDAP style queries. Currently, the module allows arbitrary queries to be performed, with the ability to list users and computers (which are implemented as LDAP filters). The new features are:
- Ability to list all domain groups
- Ability to list domain controllers
- Ability to recursively list users who belong to a specific group 
- Increased detail when listing computers

This is particularly powerful when on engagements during which it would be useful to list the effective members of a group. A large domain may have a hierarchical structure, for example:

stufus may be a member of Group A
Group A may be a member of Group B
Group B may be a member of Group C
Group C may be a member of the 'Domain Admins' group

Effectively, stufus will be a domain admin, but this will not be obvious from looking at the primary group membership, nor by looking at a flat list of users and groups. This will be even harder to spot manually on a large corporate domain with hundreds of thousands of users and groups.

Although there are plenty of alternatives, such as powershell scripts, which offer this functionality, I have submitted this PR because there is no 'native' meterpreter equivalent. I have also modified the enum_ad_users post module and added an enum_ad_groups post module for completeness.
# Usage

The following new commands have been added (accessible in meterpreter):

| Command | Description |
| --- | --- |
| adsi_dc_enum | List the domain controllers for the specified domain. The information that is provided will include the name, DNS hostname, DN, operating system, version, service pack, host description and any comments. |
| adsi_group_enum | List all groups on the specified domain. The information provided includes the name, distinguished name (DN) and description. |
| adsi_nested_group_user_enum | Recursively list all users who are effectively members of the specified group DN. |

The following POST modules have been added or modified:

| Module | Description |
| --- | --- |
| post/windows/gather/enum_ad_users | Lists AD users, with the addition of the GROUP_MEMBER module option. |
| post/windows/gather/enum_ad_groups | Lists AD groups. |
# Meterpreter example
## adsi_dc_enum

This example will list the domain controllers for the MWRDEMO domain.

```
meterpreter > adsi_dc_enum MWRDEMO

MWRDEMO Objects
===============

name    dnshostname      distinguishedname                               operatingsystem                    operatingsystemversion  operatingsystemservicepack  description  comment
----    -----------      -----------------                               ---------------                    ----------------------  --------------------------  -----------  -------
W2K3DC  w2k3dc.demo.mwr  CN=W2K3DC,OU=Domain Controllers,DC=demo,DC=mwr  Windows Server 2003                5.2 (3790)              Service Pack 1                           
W2K8DC  W2K8DC.demo.mwr  CN=W2K8DC,OU=Domain Controllers,DC=demo,DC=mwr  Windows Server 2008 R2 Enterprise  6.1 (7600)                                                       

Total objects: 2
```
## adsi_group_enum

This example will list all of the domain groups on the MWRDEMO domain.

```
meterpreter > adsi_group_enum MWRDEMO

MWRDEMO Objects
===============

name                                     distinguishedname                                                   description
----                                     -----------------                                                   -----------
Account Operators                        CN=Account Operators,CN=Builtin,DC=demo,DC=mwr                      Members can administer domain user and group accounts
Administrators                           CN=Administrators,CN=Builtin,DC=demo,DC=mwr                         Administrators have complete and unrestricted access to the computer/domain
Allowed RODC Password Replication Group  CN=Allowed RODC Password Replication Group,CN=Users,DC=demo,DC=mwr  Members in this group can have their passwords replicated to all read-only domain controllers in the domain
Backup Operators                         CN=Backup Operators,CN=Builtin,DC=demo,DC=mwr                       Backup Operators can override security restrictions for the sole purpose of backing up or restoring files
Cert Publishers                          CN=Cert Publishers,CN=Users,DC=demo,DC=mwr                          Members of this group are permitted to publish certificates to the directory
Certificate Service DCOM Access          CN=Certificate Service DCOM Access,CN=Builtin,DC=demo,DC=mwr        Members of this group are allowed to connect to Certification Authorities in the enterprise
Cryptographic Operators                  CN=Cryptographic Operators,CN=Builtin,DC=demo,DC=mwr                Members are authorized to perform cryptographic operations.
Denied RODC Password Replication Group   CN=Denied RODC Password Replication Group,CN=Users,DC=demo,DC=mwr   Members in this group cannot have their passwords replicated to any read-only domain controllers in the domain
Distributed COM Users                    CN=Distributed COM Users,CN=Builtin,DC=demo,DC=mwr                  Members are allowed to launch, activate and use Distributed COM objects on this machine.
DnsAdmins                                CN=DnsAdmins,CN=Users,DC=demo,DC=mwr                                DNS Administrators Group
DnsUpdateProxy                           CN=DnsUpdateProxy,CN=Users,DC=demo,DC=mwr                           DNS clients who are permitted to perform dynamic updates on behalf of some other clients (such as DHCP servers).
Domain Admins                            CN=Domain Admins,CN=Users,DC=demo,DC=mwr                            Designated administrators of the domain
(...snipped...)
```
## adsi_nested_group_user_enum

This example will list all of the domain groups on the MWRDEMO domain.

```
meterpreter > adsi_nested_group_user_enum MWRDEMO "CN=Domain Admins,CN=Users,DC=demo,DC=mwr"

MWRDEMO Objects
===============

samaccountname       name              distinguishedname                            description                                             comment
--------------       ----              -----------------                            -----------                                             -------
Administrator        Administrator     CN=Administrator,CN=Users,DC=demo,DC=mwr     Built-in account for administering the computer/domain  
hidden.admin         Hidden Admin      CN=Hidden Admin,CN=Users,DC=demo,DC=mwr                                                               
stuart.morgan-admin  Stuart Morgan     CN=Stuart Morgan,CN=Users,DC=demo,DC=mwr                                                          

Total objects: 3
```

In this example, the 'hidden.admin' user is not a direct member of the Domain Admins group; it is effectively a member due to the nested groups. 
# Metasploit example
## enum_ad_groups

```
use post/windows/gather/enum_ad_groups
set SESSION -1
run
```

Output is essentially the same as adsi_group_enum
## enum_ad_users

```
use post/windows/gather/enum_ad_users
set GROUP_MEMBER CN=Domain Admins,CN=Users,DC=demo,DC=mwr
set SESSION -1
run
```

Output is essentially the same as adsi_nested_group_user_enum
# Internals

Internally, all of the added functionality simply comprises LDAP filters. The ability to perform recursive user searching is implemented using the LDAP_MATCHING_RULE_IN_CHAIN matching rule OID (1.2.840.113556.1.4.1941). This is significantly more efficient than any manual recursing because, regardless of the number of nested groups, it is one query.

### CODE REVIEW

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/extapi/adsi.rb [hunk 1]
```diff
@@ -40,6 +43,52 @@ def name
...
+  # Options for the adsi_nested_group_user_enum command.
+  #
+  @@adsi_nested_group_user_enum_opts = Rex::Parser::Arguments.new(
+    "-h" => [ false, "Help banner" ],
+    "-m" => [ true, "Maximum results to return." ],
+    "-p" => [ true, "Result set page size." ]
+  )
+
+  def adsi_nested_group_user_enum_usage
+    print(
```
Comment (hmoore-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/extapi/adsi.rb]: Inline newlines can break in weird ways, please put each line into a print_line() call

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/extapi/adsi.rb [hunk 2]
```diff
@@ -81,6 +130,45 @@ def cmd_adsi_user_enum(*args)
...
+  #
+  @@adsi_group_enum_opts = Rex::Parser::Arguments.new(
+    "-h" => [ false, "Help banner" ],
+    "-m" => [ true, "Maximum results to return." ],
+    "-p" => [ true, "Result set page size." ]
+  )
+
+  def adsi_group_enum_usage
+    print(
+      "\nUsage: adsi_group_enum <domain> [-h] [-m maxresults] [-p pagesize]\n\n" +
```
Comment (hmoore-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/extapi/adsi.rb]: Same thing, use print_line() for each line of output.

**Comment by stufus** [comment]
@hmoore-r7 - I've fixed the whole adsi command dispatcher rb changing print for print_line and tidying it up, hope that's all ok

**Comment by hmoore-r7** [comment]
@stufus Thanks!

**Comment by hmoore-r7** [comment]
Looks like some errors crept in:
```
[-] WARNING! The following modules could not be loaded!
[-] 	/work/msf-hdm/modules/post/windows/gather/enum_ad_groups.rb: SyntaxError /work/msf-hdm/modules/post/windows/gather/enum_ad_groups.rb:49: syntax error, unexpected ')', expecting keyword_end
/work/msf-hdm/modules/post/windows/gather/enum_ad_groups.rb:50: dynamic constant assignment
    rescue ::RuntimeError, ::Rex::Post::Meterpreter::RequestError => e
                          ^
/work/msf-hdm/modules/post/windows/gather/enum_ad_groups.rb:50: dynamic constant assignment
    rescue ::RuntimeError, ::Rex::Post::Meterpreter::RequestError => e
                                                                    ^
/work/msf-hdm/modules/post/windows/gather/enum_ad_groups.rb:50: syntax error, unexpected =>, expecting '='
    rescue ::RuntimeError, ::Rex::Post::Meterpreter::RequestError => e
```

**Comment by stufus** [comment]
@hmoore-r7 sorry mate...umm...no idea why - it runs perfectly on mine. Let me just make sure I haven't accidentally added something and forgotten to rerun the module or something - argh...

**Comment by stufus** [comment]
@hmoore-r7 sorry mate, a rogue closing bracket made its way in there, not sure how :-(

**Comment by bcook-r7** [comment]
@hmoore-r7 did you intend to merge this? the message above says yes, but master still shows these commits are missing.

**Comment by hmoore-r7** [comment]
Strange, yes, that was intent, let me see what happened.

**Comment by hmoore-r7** [comment]
Fixed, looks like my ``git push`` ended up on my fork and not master

**Comment by stufus** [comment]
@hmoore-r7 @bcook-r7 Cheers both :-) I saw the merge to your branch hdm but I just thought that was part of your testing process or something.... but cheers though

**Pull request merged**